src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -121,99 +346,173 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
 
             if (cacheEntry is CachedResponse)
             {
-                // TODO: Compare cached request headers
-
-                // TODO: Content negotiation if there are multiple cached response formats?
-
-                // TODO: Verify content freshness, or else re-validate the data?
-
                 var cachedResponse = cacheEntry as CachedResponse;
+                var cachedResponseHeaders = new ResponseHeaders(cachedResponse.Headers);
 
-                var response = HttpContext.Response;
-                // Copy the cached status code and response headers
-                response.StatusCode = cachedResponse.StatusCode;
-                foreach (var pair in cachedResponse.Headers)
+                _responseTime = Clock.UtcNow;
+                var age = _responseTime - cachedResponse.Created;
+                age = age > TimeSpan.Zero ? age : TimeSpan.Zero;
+
+                if (EntryIsFresh(cachedResponseHeaders, age, verifyAgainstRequest: true))
                 {
-                    response.Headers[pair.Key] = pair.Value;
-                }
+                    var response = HttpContext.Response;
+                    // Copy the cached status code and response headers
+                    response.StatusCode = cachedResponse.StatusCode;
+                    foreach (var header in cachedResponse.Headers)
+                    {
+                        response.Headers.Add(header);
+                    }
 
-                // TODO: Update cache headers (Age)
-                response.Headers["Served_From_Cache"] = DateTime.Now.ToString();
+                    response.Headers[HeaderNames.Age] = age.TotalSeconds.ToString("F0", CultureInfo.InvariantCulture);
 
-                // Copy the cached response body
-                var body = cachedResponse.Body;
-                if (body.Length > 0)
+                    if (_responseType == ResponseType.HeadersOnly)
+                    {
+                        responseServed = true;
+                    }
+                    else if (_responseType == ResponseType.FullReponse)
+                    {
+                        // Copy the cached response body
+                        var body = cachedResponse.Body;
+
+                        // Add a content-length if required
+                        if (response.ContentLength == null && string.IsNullOrEmpty(response.Headers[HeaderNames.TransferEncoding]))
+                        {
+                            response.ContentLength = body.Length;
+                        }
+
+                        if (body.Length > 0)
+                        {
+                            await response.Body.WriteAsync(body, 0, body.Length);
+                        }
+
+                        responseServed = true;
+                    }
```

Comment:
I think it's sufficient to silently not serve a response when a response type is not specified. Even if we do throw an exception here, I think `InvalidOperationException` is more appropriate since the `_responseType` was `null`. The message would say something like "Response type must be set. RequestIsCacheable() must be called before trying to serve response from cache." Might also want to extract out the part inside `RequestIsCacheable` that sets this value in that case.